keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -0,0 +1,730 @@
+1. `bucketAccessName`: name of the bound cluster-scoped `BucketAccess` instance. Set by COSI.
+1. `accessGranted`: if true access has been granted to the bucket. If false then access to the bucket has not been granted.
+
+#### BucketAccess
+
+A cluster-scoped administrative custom resource which encapsulates fields from the `BucketAccessRequest` (BAR) and the `BucketAccessClass` (BAC).  The purpose of the `BucketAccess` (BA) is to serve as an access communication path between provisioners and the central COSI controller.  The COSI controller creates a `BucketAccess` instance for a new `BucketAccessRequest`, and there is a 1:1 mapping between `BucketAccess` and `BucketAccessRequest`.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketAccess
+metadata: 
+  name: [1]
+  labels:
+    cosi.io/provisioner: [2]
+  finalizers:
+  - cosi.io/finalizer [3]
+ spec:
+  provisioner: [4]
+  bucketInstanceName: [5]
+  bucketAccessRequest: [6]
+  serviceAccount: [7]
+  mintedSecretName: [8]
+  policyActionsConfigMapData: [9]
+  principal: [10]
+  parameters: [11]
+ status:
+   accessGranted: [12]
+```
+
+1. `name`: the BA name is generated by COSI in this format: _"<BAR.namespace*>-<BAR.name*>-<uuid>"_. The uuid is unique within a cluster. `*` the first 100 characters of the namespace and the name are used.
+1. `labels`: added COSI.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+1. `finalizers`: added by COSI to defer `BucketAccess` deletion until the pod that uses this access terminates, and the related `BucketAccessRequest` is deleted.
+1. `provisioner`: (optional) The provisioner field as defined in the `BucketAccessClass`. If empty then there is no driver/sidecar, thus the admin had to create the `Bucket` and `BucketAccess` instances.
+1. `bucketInstanceName`:  name of the `Bucket` instance bound to this BA.
+1. `bucketAccessRequest`: an `objectReference` containing the name, namespace and UID of the associated `BucketAccessRequest`.
+1. `serviceAccount`: an `objectReference` containing the name, namespace and UID of the associated `ServiceAccount`. Empty when the `BucketAccessRequest.mintedSecretName` is specified.
+1. `mintedSecretName`: name of the admin created or provisioner-generated Secret containing access credentials. This Secret exists in the provisioner’s namespace and must be copied to the app namespace by the COSI controller. **Note:** the provisioner's namespace is contained in the registered driver name.
+1. `policyActionsConfigMapData`: encoded data that contains a set of provisioner/platform defined policy actions to a given user identity. Contents of the ConfigMap that the *policyActionsConfigMap* field in the `BucketAccessClass` refers to.
```

Comment:
What does the content of the `policyActionsConfigMap` look like? Is it defined by the driver? If so, can you provide an example of the type of information it would be used to track. What happens if the object doesn't exist? What happens if it exists, but is then deleted?

Reply:
BA.policyActionsConfigMapData could look like this:
```
{“Effect”:“Allow”,“Action”:“s3:PutObject”,“Resource”:“arn:aws:s3:::profilepics/*“},
{“Effect”:“Allow”,“Action”:“s3:GetObject”,“Resource”:“arn:aws:s3:::profilepics/*“},
{“Effect”:“Deny”,“Action”:“s3:*“,”NotResource”:“arn:aws:s3:::profilepics/*“}
```
It’s content is defined by the driver but is created by the user in their ns. The CM content is copied into the BA as the _source of truth_ to handle cases where the CM is replaced, and to allow the sidecar to remain minimally privileged as to not need access to the user's ns. If the CM does not exists when creating the BA we would retry (wait and requeue).

Reply:
Just to clarify do `Principal` allowed here, or admin can only specify `resources`?

Reply:
In the workflow of how we provisioner BA, the principal is not known upfront. The policy should only be about the resources themselves, and when the access is minted, the policy is applied to the minted prinicipal